packages/backend/server/src/models/workspace-user.ts

```diff
@@ -0,0 +1,385 @@
+    }
+  }
+
+  /**
+   * Set or update the Role of a user in a workspace.
+   *
+   * NOTE: do not use this method to set the [Owner] of a workspace. Use {@link setOwner} instead.
+   */
+  @Transactional()
+  async set(
+    workspaceId: string,
+    userId: string,
+    role: WorkspaceRole,
+    defaultStatus: WorkspaceMemberStatus = WorkspaceMemberStatus.Pending
+  ) {
+    if (role === WorkspaceRole.Owner) {
+      throw new Error('Cannot grant Owner role of a workspace to a user.');
+    }
+
+    const oldRole = await this.get(workspaceId, userId);
+
+    if (oldRole) {
+      if (oldRole.type === role) {
+        return oldRole;
+      }
+
+      const newRole = await this.db.workspaceUserRole.update({
+        where: { id: oldRole.id },
+        data: { type: role },
+      });
+
+      if (oldRole.status === WorkspaceMemberStatus.Accepted) {
+        this.event.emit('workspace.members.roleChanged', {
+          userId,
+          workspaceId,
+          role: newRole.type,
+        });
+      }
+
+      return newRole;
+    } else {
+      return await this.db.workspaceUserRole.create({
+        data: {
+          workspaceId,
+          userId,
+          type: role,
+          status: defaultStatus,
+        },
+      });
+    }
+  }
+
+  async setStatus(
+    workspaceId: string,
+    userId: string,
+    status: WorkspaceMemberStatus
+  ) {
+    return await this.db.workspaceUserRole.update({
+      where: {
+        workspaceId_userId: {
+          workspaceId,
+          userId,
+        },
+      },
+      data: {
+        status,
+      },
+    });
+  }
+
+  async accept(id: string) {
+    await this.db.workspaceUserRole.update({
+      where: { id },
+      data: { status: WorkspaceMemberStatus.Accepted },
+    });
+  }
+
+  async delete(workspaceId: string, userId: string) {
+    await this.db.workspaceUserRole.deleteMany({
+      where: {
+        workspaceId,
+        userId,
+      },
+    });
+  }
+
+  async get(workspaceId: string, userId: string) {
+    return await this.db.workspaceUserRole.findUnique({
+      where: {
+        workspaceId_userId: {
+          workspaceId,
+          userId,
+        },
+      },
+    });
+  }
+
+  async getById(id: string) {
+    return await this.db.workspaceUserRole.findUnique({
+      where: { id },
+    });
+  }
+
+  /**
+   * Get the **accepted** Role of a user in a workspace.
+   */
+  async getActive(workspaceId: string, userId: string) {
+    return await this.db.workspaceUserRole.findUnique({
+      where: {
+        workspaceId_userId: { workspaceId, userId },
+        status: WorkspaceMemberStatus.Accepted,
+      },
+    });
+  }
+
+  async getOwner(workspaceId: string) {
+    const role = await this.db.workspaceUserRole.findFirst({
+      include: {
+        user: true,
+      },
+      where: {
+        workspaceId,
+        type: WorkspaceRole.Owner,
+      },
+    });
+
+    if (!role) {
+      throw new Error('Workspace owner not found');
```

Comment:
If such a data anomaly occurs, will it keep reporting errors and then use alerts for manual data correction?

Reply:
it only happens when user deleted but owned workspaces has not been deleted yet. no fixes can be applied.